fix byobu, closes #14443

ryantm opened the pull request:
###### Motivation for this change

byobu-config didn't work (issue #14443) ; byobu is out of date.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

ryantm committed:
byobu: 5.112 -> 5.115; closes #14443

This adds a bunch of missing dependencies to byobu, including libnewt's
snack python module. byobu-config now works correctly, closing #14443.

globin commented:
@FRidh does that python wrapping look right?

FRidh commented:
@globin no.

I think we already have newt, pkgs/development/libraries/newt/default.nix

FRidh commented:
I just had a closer look at it. All you seem to need is the Python interpreter. The shebangs of scripts will automatically be updated to point to the interpreter. Why do you prefix PYTHONPATH? What is needed for?

ryantm commented:
Thanks for pointing out the other copy of newt. I'll fix that and try to fix the python wrapping part.

@FRidh Why do you prefix PYTHONPATH? What is needed for?

libnewt provides a python module called "snack". byobuy-config depends on that.

ryantm commented:
I updated it to use the existing newt library. 

@FRidh I tried to figure out how to use python.withPackages but it seems to be for python modules that exist in the standard python module architecture. If you can tell me how to use it correctly in this case, I'll be happy to update the PR. I changed the PYTHONPATH to be set instead of prefixed in case python.withPackages cannot be used.

ryantm commented:
@FRidh bump. Are you okay with how I set the PYTHONPATH?

I think this PR is ready to merge.

FRidh commented:
@ryantm thanks for the reminder. I am surprised `python.withPackages(ps: [newt])` doesn't function and will look at that later. What you've done now seems okay to me.

There is a downside of adding `python` here: packages that have a runtime dependency on `newt` will now also have `python` in their closure. I suppose programs using `newt` don't use any of the scripts, and thus also not snack/python, right? If that's the case, then what we typically do is split the library in multiple outputs.

FRidh commented:
I'm testing now a modification where I've created a separate output for snack.

globin commented:
Could you drop the libnewt init and revert commits to cleanup the commit history?

ryantm commented:
@globin sure, I'll clean it up after we finish making it work well.

@FRidh I tried using your `newt` branch. It doesn't seem to work with byobu-config and I think I know why roughly:

```
 ryantm@home1 ~/p/nixpkgs (byobu)$ NIX_PATH=nixpkgs=. nix-shell --pure -p byobu                                                                                                                                     

[nix-shell:~/p/nixpkgs]$ byobu -v
/nix/store/g896qmsijwdg56046kbwxd5spa9j4j6g-byobu-5.115/bin/byobu: line 49: locale: command not found
byobu version 5.115
tmux: need UTF-8 locale (LC_CTYPE) but have ANSI_X3.4-1968

[nix-shell:~/p/nixpkgs]$ byobu-config

[nix-shell:~/p/nixpkgs]$ exit
ryantm@home1 ~/p/nixpkgs (byobu)$ NIX_PATH=nixpkgs=. nix-shell  -p byobu

[nix-shell:~/p/nixpkgs]$ byobu-config
ERROR: Could not import the python snack module

[nix-shell:~/p/nixpkgs]$ cat `which byobu-config`
#!/nix/store/b3hnjqy2py1hbh364i6jj6jmh9ajm79l-bash-4.4-p5/bin/sh -e
#
#    byobu-config - interactive byobu configuration wrapper script
#    Copyright (C) 2013-2014 Dustin Kirkland
#
#    Authors: Dustin Kirkland <kirkland@byobu.org>
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, version 3 of the License.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

PKG="byobu"
[ -z "${BYOBU_PREFIX}" ] && export BYOBU_PREFIX="/nix/store/g896qmsijwdg56046kbwxd5spa9j4j6g-byobu-5.115" || export BYOBU_PREFIX
. "${BYOBU_PREFIX}/lib/${PKG}/include/common"

${BYOBU_PYTHON} "${BYOBU_PREFIX}/lib/${PKG}/include/config.py"
```

In an impure shell, byobu-config seems to be discovering my system python install and snack isn't there of course. When I use a pure shell, it finds the correct python environment and works! (Maybe this is why python.withPackages (ps: [newt]) didn't work for us before?)

FRidh commented:
What is `${BYOBU_PYTHON}` , where is it set? If it hasn't been fixed during building it will remain open for impurities. Its funny actually, wrapping a wrapper :)

ryantm commented:
@Fridh So, BYOBU_PYTHON ends up being blank because the /lib/include/constants.py fails to calculate a valid BYOBU_PYTHON, but the config.py file above has a python in its shebang.

If we just set BYOBU_PYTHON, that stops it from trying to auto detect python in an impure way.

I've tested it now and byobu-config works in a pure or impure shell.

I've squashed all the commits together, and tested this commit some. I hope you don't mind combining your commits into mine like that. If you do I can try to fix it up in a more careful way.

FRidh commented on pkgs/development/libraries/newt/default.nix:
```diff
@@ -1,31 +1,48 @@
+    sha256 = "101fzx5n711wj01rpkcixyfc1iklny8r9fdsgimaz5hrq9bdph08";
   };
 
+  postUnpack = ''
```
this should only be done with `pythonSupport`

FRidh reviewed (changes requested):
> If we just set BYOBU_PYTHON, that stops it from trying to auto detect python in an impure way.

Good solution!

> I've squashed all the commits together, and tested this commit some. 

Please separate it into two commits, one for byobu, and another for newt explaining the version bump, and multiple outputs with snack going into `python-packages.nix`.

> I hope you don't mind combining your commits into mine like that.

No, that's no problem.

ryantm commented:
I think there are some problems with this pull request as is. For example, it seems like it causes byobu-launch to crash the shell for me. I need to look into it more.

FRidh commented:
Any update?

ryantm commented:
I'm going to close this, because I stopped using byobu rather than debug all the ways it does impure stuff. I just use tmux directly.
